### ISSUE TITLE
Issue with Serverless / Api Gateway / ServiceStack combination request deserialization

### ISSUE THREAD

**Issue by ryanhelms** [ServiceStack/Issues#599]
Hey mythz,

First, I have to say that this combination is very powerful when it works properly. We are able to have infinitely scalable, high-functioning ServiceStack endpoint that are just as feature rich as any other  ServiceStack endpoint deployed in a container / EC2 environment. 

We have endpoints interfacing with Virtual File Systems, MQ pub/sub, etc but oddly enough, we break down when it comes to executing a POST / PUT to an endpoint and  expecting to get a deserialized request object. The object basically comes through as a new object of the request type (empty - `Payload` property in the example below) with any properties that I explicitly define in the return coming back as expected (`Ping` and `UpTime` properties in the example below). I have setup a demo solution [here](https://github.com/ryanhelms/servicestack-serverless-apigateway) for you to have a look at.

**Note**: I have an example api deployed using the steps below. You can use this one if you'd like but I was verbose here in case you wanted to reproduce it yourself from the ground up. 

Example endpoint deployed from steps below: 
 - https://a0pxyby83h.execute-api.us-east-2.amazonaws.com/demo/ping

===================================================================

The steps to reproduce the problem are:

1.  Clone the [repo](https://github.com/ryanhelms/servicestack-serverless-apigateway)

2.  Open your command prompt

3. `cd` to MyApp directory

4.  `npm install` 

5. Edit MyApp -> serverless.yml
    - change `<AWS_ROLE>` to an existing role
    - change `<SERVERLESS_DEPLOYMENT_BUCKET>` to a temporary bucket to put serverless deployments

6. From your command prompt, execute: `sls deploy --stage <anything> --region <valid_region>` 

7. After a successful deployment, get the endpoint url that was assigned for your service:
    example: https://a0pxyby83h.execute-api.us-east-2.amazonaws.com/demo/ping

8. Next, execute a GET request:
    - You should get back a response similar to:
>        {
>            "ping": "Get: Pong",
>            "upTime": "Service has been up for -XX:XX:XX.XXXXXXXX"
>        }`

9.  Next, execute a POST request with:
    - the `Accept` header set to `application/json`
    - the `Content-Type` header set to `application/x-www-form-urlencoded`
    - the request property `Payload`  set to `payload data`

    - You should get back a response similar to:
>        {
>            "payload": "payload data",
>            "ping": "Post: Pong",
>            "upTime": "Service has been up for -XX:XX:XX.XXXXXXXX"
>        }`

10.  Next, execute a POST request with:
    - the `Accept` header set to `application/json`
    - the `Content-Type` header set to `application/json`
    - the request property `Payload`  set to `payload data`

    - You should get back a response similar to:
>        {
>            "ping": "Post: Pong",
>            "upTime": "Service has been up for -XX:XX:XX.XXXXXXXX"
>        }`

Obviously, this is where the problem is. I expect to get back `payload: "payload data"`  here just as I did for the previous post.

Any ideas are welcome. I'd love to get this working 100% as it has huge potential.

Cheers!

**Comment by Layoric** [comment]
@ryanhelms if you have API Gateway in the mix, I suggest double checking any body mapping templates as if none are setup, API gateway will not pass them on. 

https://docs.aws.amazon.com/apigateway/latest/developerguide/request-response-data-mappings.html

**Comment by mythz** [comment]
@ryanhelms If you still have issues can you [register the RequestLogger](http://docs.servicestack.net/request-logger) to see the Request that's sent to ServiceStack.

**Comment by ryanhelms** [comment]
Thanks @layoric / @mythz, I'll update you soon.

**Comment by ryanhelms** [comment]
Hey guys, I've made progress with this but we are still having few issues. 

First, we have the request coming through as:

     {
        "resource": "/ping",
        "path": "/ping",
        "httpMethod": "POST",
        "headers": {
            "Accept": "application/json",
            "Accept-Encoding": "gzip, deflate",
            "cache-control": "no-cache",
            "CloudFront-Forwarded-Proto": "https",
            "CloudFront-Is-Desktop-Viewer": "true",
            "CloudFront-Is-Mobile-Viewer": "false",
            "CloudFront-Is-SmartTV-Viewer": "false",
            "CloudFront-Is-Tablet-Viewer": "false",
            "CloudFront-Viewer-Country": "US",
            "Content-Type": "application/json",
            "Host": "z2s930hre6.execute-api.us-east-2.amazonaws.com",
            "Postman-Token": "4e3f7a9b-0b0c-48f2-a391-e9dc311127b9",
            "User-Agent": "PostmanRuntime/7.1.1",
            "Via": "1.1 73e977a766ff8fd41fbb5fef8e1f4f2e.cloudfront.net (CloudFront)",
            "X-Amz-Cf-Id": "4aCnBt0l19fvITYuvsBYpKCfNujWyJs6ueHIHw4MPW3pr27RSfUhAg==",
            "X-Amzn-Trace-Id": "Root=1-5a88c99f-a43a3b889c037e545eab745e",
            "X-Forwarded-For": "216.196.242.118, 52.46.22.85",
            "X-Forwarded-Port": "443",
            "X-Forwarded-Proto": "https"
        },
        "requestContext": {
            "path": "/demo/ping",
            "accountId": "213343987366",
            "resourceId": "0s94gq",
            "stage": "demo",
            "requestId": "34836235-1443-11e8-827f-7911bb1c4ea4",
            "identity": {
                "sourceIp": "216.196.242.118",
                "userAgent": "PostmanRuntime/7.1.1"
            },
            "resourcePath": "/ping",
            "httpMethod": "POST",
            "apiId": "z2s930hre6"
        },
        "body": "{\"Payload\": \"some data\"}",
        "isBase64Encoded": false
    }

Next, we exposed two properties, `Request` and `RawRequest` on the `PingResponse`, as follows:

    public class PingResponse
    {
        /// <summary>
        /// Temporary: used to display the request passed in from the ServiceStack method
        /// </summary>
        public object Request { get; set; }

        /// <summary>
        /// Temporary: used to display the raw request passed in from the ServiceStack service
        /// </summary>
        public object RawRequest { get; set; }
    }

Next, we executed the following request:

    POST https://z2s930hre6.execute-api.us-east-2.amazonaws.com/demo/ping
    { 
        "Payload": "some data"
    }

The response to this request would be expected to look like:
    
    {
        "request": {
            "payload": "some data"
        },
        "rawRequest": {
            "payload": "some data"
        }
    }

but in fact, it looks like:

    {
        "request": { },
        "rawRequest": {
            "payload": "some data"
        }
    }

This leads me to my question. If the following works:

    Request.GetRawBody().FromJson<PingRequest>() 

why is the request that is passed into the `Post(object request)` method `null`?

**Comment by mythz** [comment]
The Request DTO must be a concrete Type which specifies which Type the Response Body deserializes into. 

If you want to accept Raw Data you can either have your Request DTO [implement IRequiresRequestStream](http://docs.servicestack.net/serialization-deserialization#reading-directly-from-the-request-stream) to access the raw byes from the Request Stream or if you control the client have it send raw data in a `string` or `byte[]` Request DTO property.

Closing this as late-bound `object` Request DTOs or properties are not supported. For untyped request bodies you can use a `Dictionary<string,string>` if it's flat JSON object, or send serialized JSON in a `string` property if it's not.

**Comment by ryanhelms** [comment]
@mythz , I mis-typed above. the Post signature is `Post(PingRequest request)`.

**Comment by mythz** [comment]
Ok but my comment still applies if you're trying to accept unknown payloads. The `object` properties is only for deserializing types where the Concrete .NET Type is specified in the payload with `__type` property (which is emitted by ServiceStack.Text when serializing object or interface properties).

**Comment by mythz** [comment]
What's the class definition of `PingRequest` and by doesn't work do you mean it's not serialized? Can you confirm that it doesn't work locally?

Request DTOs are never `null`, if no data was sent an Empty Request DTO is passed to the Service.

**Comment by ryanhelms** [comment]
The code for the entire solution is [here](https://github.com/ryanhelms/servicestack-serverless-apigateway/). The `PingRequest` DTO is located [here](https://github.com/ryanhelms/servicestack-serverless-apigateway/blob/master/MyApp.ServiceModel/PingRequest.cs#L11). The problem is, when you run this locally, the results are not the same as when you run it behind API Gateway. 

This is a perfect example of what is happening:

If you run the `Can_call_Ping_Service_Via_Post` from the `MyApp.Tests` project, you will receive a test pass. It is performing the following:

        [Test]
        public void Can_call_Ping_Service_Via_Post()
        {
            var client = CreateClient();
            var request = new PingRequest {Payload = "Some data"};
            var response = client.Post(request);

            Assert.That(response.Ping, Contains.Substring(request.ToJson()));
        }

The `response` from the request in the test above looks like:

    {
        "ping": "Post: Pong => Request: {\"payload\":\"Some data\"}",
        "upTime": "Service has been up for 00:00:04.0002838"
    }

Everything is as expected at this point. 

If I deploy this and execute the same request:

    POST https://z2s930hre6.execute-api.us-east-2.amazonaws.com/demo/ping
    {
        "Payload": "some data"
    }

I get the following response:

    {
        "ping": "Post: Pong => Request: {}",
        "upTime": "Service has been up for 00:00:04.0002838"
    }

With that being said, here is a copy of the `APIGatewayProxyRequest` that is sent in as the `request` parameter from lambda:

    {
        "resource": "/ping",
        "path": "/ping",
        "httpMethod": "POST",
        "headers": {
            "Accept": "application/json",
            "Accept-Encoding": "gzip, deflate",
            "cache-control": "no-cache",
            "CloudFront-Forwarded-Proto": "https",
            "CloudFront-Is-Desktop-Viewer": "true",
            "CloudFront-Is-Mobile-Viewer": "false",
            "CloudFront-Is-SmartTV-Viewer": "false",
            "CloudFront-Is-Tablet-Viewer": "false",
            "CloudFront-Viewer-Country": "US",
            "Content-Type": "application/json",
            "Host": "z2s930hre6.execute-api.us-east-2.amazonaws.com",
            "Postman-Token": "682a86c8-27db-44c9-a288-3c12452ad366",
            "User-Agent": "PostmanRuntime/7.1.1",
            "Via": "1.1 5a87bdc6ee46b099c44b6807f0565eda.cloudfront.net (CloudFront)",
            "X-Amz-Cf-Id": "8iIsiff7V-2NlA2qjL_7Hzvg1pxVlQgt2ihObcFXbeAduELrbpcRVQ==",
            "X-Amzn-Trace-Id": "Root=1-5a8b7992-38152c081e22b95c196cda50",
            "X-Forwarded-For": "216.196.242.118, 52.46.18.59",
            "X-Forwarded-Port": "443",
            "X-Forwarded-Proto": "https"
        },
        "requestContext": {
            "path": "/demo/ping",
            "accountId": "213343987366",
            "resourceId": "0s94gq",
            "stage": "demo",
            "requestId": "41ad9262-15dd-11e8-bcf4-53f11af5404e",
            "identity": {
                "sourceIp": "216.196.242.118",
                "userAgent": "PostmanRuntime/7.1.1"
            },
            "resourcePath": "/ping",
            "httpMethod": "POST",
            "apiId": "z2s930hre6"
        },
        "body": "{\n\t\"Payload\": \"some data\"\n}\n\n",
        "isBase64Encoded": false
    }

As you can see, the body property which is what should be deserialized into the `PingRequest`  is not being deserialized. 

Also, as I stated earlier in the thread if I execute:

    request =  Request.GetRawBody().FromJson<PingRequest>();

as the first line of my service, everything acts as is normally should. This will end up being our work around, temporarily, to be able to use ServiceStack endpoints if we have to but I would hate to muddy up the services with the extra line 'just because'. 

Thanks for your extended amount of time on this, it is appreciated. I'm doing my best to be thorough in my explanation and to give reproducible code / examples.

**Comment by ryanhelms** [comment]
Also, for your viewing pleasure, I've done the `GetRawBody()`  work-around that I referenced above to the PUT method. 

Correct response: [source](https://github.com/ryanhelms/servicestack-serverless-apigateway/blob/master/MyApp.ServiceInterface/PingService.cs#L61)

    PUT https://z2s930hre6.execute-api.us-east-2.amazonaws.com/demo/ping
    {
        "Payload": "some data"
    }

Incorrect response:  [source](https://github.com/ryanhelms/servicestack-serverless-apigateway/blob/master/MyApp.ServiceInterface/PingService.cs#L81)

    POST https://z2s930hre6.execute-api.us-east-2.amazonaws.com/demo/ping
    {
        "Payload": "some data"
    }

**Comment by Layoric** [comment]
@ryanhelms noticing that your payload is using title case `Payload` vs camel case `payload`. I would double check your mappings as they are also case sensitive. 
Any chances you can share them?

**Comment by mythz** [comment]
There isn't different behavior behind an API gateway, it's just not being simulated correctly, i.e. you can't send raw JSON into a C# `string` property, i.e. it needs to be an escaped JSON string, e.g:

in C# it would look like:

```csharp
new PingRequest {
    Payload = JsonSerializer.SerializeToString(payload)
}
```

In JavaScript:

```js
JSON.stringify({ Payload: JSON.stringify(payload) })
```

In an escaped C# String it would look something like:

    "body": "{\n\t\"Payload\":\\\"some data\\\"\n}\n\n"",


With that said I've just added support for deserializing arbitrary JSON and JS literals into `object` properties which deserializes into untyped .NET Data structures as [seen in these tests](https://github.com/ServiceStack/ServiceStack.Text/blob/2a69915496569d8686385036374596a00872b7a3/tests/ServiceStack.Text.Tests/RuntimeSerializtionTests.cs#L232-L312).

Which is enabled by default in the latest v5.0.3 that's now [available on MyGet](http://docs.servicestack.net/myget).

Basically your Request DTO can now look like:

```csharp
[Route("/ping")]
public class PingRequest : IReturn<PingResponse>
{
    public object Payload { get; set; }
}
```

and Payload will be populated into native .NET data structures, e.g. `Dictionary<string,object>` if payload was a JSON object literal .

**Comment by ryanhelms** [comment]
@mythz I still don't understand why the request from above does not get deserialized into the request parameter in my `Post` method but calling `request = Request.GetRawBody().FromJson<PingRequest>()` seems to deserialize it just fine. This is the only thing I am trying to do. I just want the request to come through with the data from my Post body. Payload is just a string, not an object or anything complex (and intended as a simple example). I would gladly pay whatever hourly support cost that there might be to get this resolved. Is that even an option? Apparently I am either too dense or have been staring at the issue too long to get past it.

@Layoric , I am currently use the lambda proxy integration. I attempted to use the request mappings but it seems that was a lot of undue overhead for what is trying to be done. In an effort to keep things as lean as possible, it is hard to believe passing the post body would require these mappings. 

In the event that you say, "quit being stupid, it does", do you have any examples of how you would do it? I had a response template that looked like [this](https://github.com/ryanhelms/servicestack-serverless-apigateway/blob/master/serverless/requestTemplates/request.tmpl.json) while doing some testing but after an hour with AWS support and looking at what we were trying to do, they insisted that my body was being passed through properly to be accepted by lambda, especially since I was getting a 200 from the lambda response. 

I still have faith ......

**Comment by mythz** [comment]
@ryanhelms My previous comment specifies how to [get the raw JSON string](https://github.com/ServiceStack/Issues/issues/599#issuecomment-366486458) by having your Request DTO implement `IRequiresRequestStream` which is similar to `Request.GetRawBody()` but ServiceStack will skip deserializing the Request Body.

So if I just needed the raw JSON body sent, my Request DTO would look like:

```csharp
public class PingRequest : IRequiresRequestStream, IReturn<PingResponse>
{
    public Stream RequestStream { get; set; }
}
```

and my Service impl would be:

```csharp
public object Post(PingRequest request)
{
    string json = request.RequestStream.ReadFully().FromUtf8Bytes();
}
```

It would help if you could provide the **exact JSON** that's sent, i.e. the `json` contents saved to a `file.json`.

**Comment by ryanhelms** [comment]
@mythz I don't want to do anything with raw json or raw anything at all. I just want this to work the way that every other ServiceStack service that I own works. :)

Here is the request that's being sent in: [file.json](https://github.com/ServiceStack/Issues/files/1740390/file.json.txt)

Here is a screenshot of my postman client: ![postman client](https://user-images.githubusercontent.com/1596639/36428934-3d14d604-161f-11e8-8c21-548cca6dad19.png)

**Comment by ryanhelms** [comment]
@mythz my first post in this thread is an exact replication of the issue all the way from the clone of the deployment of the serverless lambda endpoint. The only thing is does not have is the clean up method. In order to clean up the endpoints:

- `cd MyApp`
- `sls remove --stage demo --region us-east-2`

Once that is done, no resources will remain on the account and you can remove any roles, s3 buckets or other resources that were created manually.

**Comment by ryanhelms** [comment]
Also, as for your need for a locally reproducible test, I think this particular situation might fall outside of that. If you run the `MyApp.Tests.IntegrationTests.Can_call_Ping_Service_Via_Post()`  test, you will see that it passes as expected. If you execute that same test via the deployed lambda endpoint, you'll see that the request parameter basically comes in as a new DTO of PingRequest, rather than a DTO with data, i.e: `Payload` having a value of `some data`. I hope this clarifies it further.

**Comment by ryanhelms** [comment]
Knowing that you haven't dabbled with ServiceStack running on a lambda endpoint, this should be an exciting addition. It would be an absolute amazing selling point to any AWS user to be able to deploy these  endpoints in a managed environment. Hopefully you see my goal.

**Comment by ryanhelms** [comment]
Also, if you open a new cmd prompt and:

 - `cd MyApp`
 - `sls logs --stage demo --region us-east-2` -f Ping

this will stream the logs for that function to your command prompt

**Comment by mythz** [comment]
ok this took me way too long to diagnose, the project had missing deps that weren't on NuGet, had no idea serverless/sls wasn't an AWS cli tool, the "with appropriate lambda deployment permissions" step took me over an hour to play whack-a-mole, which led to corrupted cloud formation template that took me a while to figure out I had to manually delete. This environment is also painful to debug resorting to printf debugging and slow deploys/iteration times. 

Anyway the issue should be fixed with [this commit](https://github.com/ServiceStack/ServiceStack/commit/a4543fa9540c5fe524a35ce141702f28160555a3) basically .NET Core was reporting the RequestBody had ContentLength=0 and no Content-Encoding heuristic , but turns we can use the Request InputStream Length to determine if the Request has a request body. 

Note: by locally reproducible test I mean a test I can actually debug, e.g. a local test calling a 3rd Party API is not useful unless the issue is with the client library, but I can see how it's not possible here for simulating AWS API Gateway / AWS Lambda.

Given how it takes so long to debug AWS API Gateway interoperability issues, future issues that can't be simulated and debugged locally will have lower priority support, postponed after the reserved support down time allocated immediately after each release.

**Comment by ryanhelms** [comment]
Very sorry. I even opened up the nuget server for you but forgot to add that in the nuget.config for you :\ I apologize. I appreciate you digging into this and resolving the issue, it really helped me through a situation ? 
Thanks again.
